### PR TITLE
PSA: Update AEAD output buffer macros to PSA API version 1.0

### LINKED ISSUE #4163: Non-compliance of PSA AEAD output size macros
From the [changes between version 1.0beta3 and 1.0.0](https://armmbed.github.io/mbed-crypto/html/appendix/history.html#changes-between-1-0-beta-3-and-1-0-0):

> AEAD output buffer size macros are now parameterized on the key type as well as the algorithm:
> * `PSA_AEAD_ENCRYPT_OUTPUT_SIZE()`
> * `PSA_AEAD_DECRYPT_OUTPUT_SIZE()`
> * `PSA_AEAD_UPDATE_OUTPUT_SIZE()`
> * `PSA_AEAD_FINISH_OUTPUT_SIZE()`
> * `PSA_AEAD_TAG_LENGTH()`
> * `PSA_AEAD_VERIFY_OUTPUT_SIZE()`

Unfortunately these macros already exist under the same name in Mbed TLS without the key type argument. So we can't be compliant to the PSA API without breaking the Mbed TLS API.

My initial proposal is to fix this in the 3.0 branch. To be discussed.

### PR DESCRIPTION
**Pull request by bensze01**
## Description
The PSA Crypto API version 1.0 added extra parameters to the AEAD output buffer macros.
This PR updates them, and provides backwards compatibility macros using C99 vararg macros.

Closes #4163.

## Status
**READY**

## Requires Backporting
When there is a bug fix, it should be backported to all maintained and supported branches.
Changes do not have to be backported if:
No, PSA only.

## Migrations
Yes, to bring us in line with PSA Crypto version 1.0.
Compatibility macros are provided.

## Todos
- [x] Update documentation
- [x] Update uses of newly deprecated macros
- [x] Update Changelog
- [ ] Forward-port for development

### CODE REVIEW

**Comment by bensze01** [comment]
The Travis failure is due to a bug on the development branch, that has since been fixed.

I cannot seem to re-trigger it without pushing new changes, so I'm commenting to let everyone know it can be safely ignored.

**Review by gabor-mezei-arm — APPROVED** [review]
LGTM

**Comment by bensze01** [comment]
Rebased PR to fix merge conflicts.

**Review by gabor-mezei-arm — DISMISSED** [review]

**Review by ronald-cron-arm — COMMENTED** [review]
I think it is fair to say about the variadic macros introduced by this PR that:
1) they are complex and thus will be difficult to maintain. They are made of a lot of macros. Their usage results in macro expansions of unreasonable length (see  "Stop macro expansion in ASSERT_EQUAL error message).
2) given that we already need some compiler specific and even compiler version specific code to make them work with the compilers we support (type and version) it is highly likely that they won't work with some other compilers.

Thus are we sure we need this ? I understand that this is done to still support the old AEAD output length macros  with the same name but a different number of parameters. So far, we've been able to maintain the compatibility with the macros and functions of PSA v1-beta that didn't make it to v1-release. But wouldn't it be acceptable to say here that the cost/risk of maintaining the compatibility is too high and thus we don't do it and provide a migration path instead?

**Comment by bensze01** [comment]
@ronald-cron-arm
1. While the variadic compatibility macros contribute slightly to the overly long expansion problem, the actual root cause is in the 1.0 version of the macros introduced in the patch. In particular, the macros use `PSA_AEAD_NONCE_LENGTH` and `PSA_AEAD_TAG_LENGTH` to validate their arguments, the latter of which also calls `PSA_AEAD_NONCE_LENGTH`, which in turn uses `PSA_ALG_AEAD_WITH_DEFAULT_LENGTH_TAG` three times. So the first point would be a problem even without the compatibility layer.

2. The only compiler specific code I remember having introduced is the local suppression of spurious MSVC warning about standards conformant C code. It does not depend on the specific version of MSVC either. You could argue that `MBEDTLS_EXPAND` is also compiler specific, since it is only needed for MSVC support, but using it  doesn't use Ifdefs. Any other commits referencing other compilers were instances where the C code was not actually standards conformant, which had to be corrected.

As for the maintenance burden, I only intended this patch for the 2.x branch, with the intention being we simply break the API with the first 3.0 release. I've also tried to layer the macros logically, and in a way that makes it unnecessary to delve deeper into the inner workings of the compatibility layer to update the macros' implementation, as long as we don't change the number of arguments they take again in the future - which we Gilles has said we will avoid, and shouldn't affect the 2.x branch anyhow, with development moving on to 3.0

**Comment by ronald-cron-arm** [comment]
>     1. While the variadic compatibility macros contribute slightly to the overly long expansion problem, the actual root cause is in the 1.0 version of the macros introduced in the patch. In particular, the macros use `PSA_AEAD_NONCE_LENGTH` and `PSA_AEAD_TAG_LENGTH` to validate their arguments, the latter of which also calls `PSA_AEAD_NONCE_LENGTH`, which in turn uses `PSA_ALG_AEAD_WITH_DEFAULT_LENGTH_TAG` three times. So the first point would be a problem even without the compatibility layer.
> 
ok then there is maybe something to do here. Those are macros aimed to be used by applications. We have encountered problems to use them because their expansion is very large thus some applications are likely to hit the same kind of problems at some point. One solution I see is to implement them as functions. We have psa_crypto_client.c to contain such functions needed on the client side.

**Comment by bensze01** [comment]
We promise in the API docs, that these macros will return compile-time constants - so I don't think we can implement them as functions.

I think I can modify PSA_AEAD_NONCE_LENGTH to make this problem go away, for now at least - even with the compat layer applied.

**Comment by ronald-cron-arm** [comment]
> 
> 
> We promise in the API docs, that these macros will return compile-time constants - so I don't think we can implement them as functions.
> 
Right, forgot about that.

**Comment by bensze01** [comment]
I've removed the macro expansion length workaround, and pushed a new commit that reduces the expanded size of the macros considerably.

**Comment by ronald-cron-arm** [comment]
Good that is a much better solution to me. What is the maximum expanded size now?

**Comment by bensze01** [comment]
The longest macro expands to 2343 characters, or 2119 characters with `MBEDTLS_DEPRECATED_REMOVED` (the compat layer adds 224 characters).

This is down from 6482 and 6259 before the latest commit.

**Comment by ronald-cron-arm** [comment]
> 
> 
> The longest macro expands to 2343 characters, or 2119 characters with `MBEDTLS_DEPRECATED_REMOVED` (i.e. the compat layer adds 224 characters).
> 
> This is down from 6482 and 6259 before the latest commit.

Great, that's good improvement. Any other idea to reduce this further as over 2000 characters is still quite big ?

**Comment by bensze01** [comment]
With the commit I just pushed, I've further reduced the expanded size of the longest macro to 1167 characters, or 1055 with `MBEDTLS_DEPRECATED_REMOVED` defined.

File: include/psa/crypto_compat.h
Comment (gilles-peskine-arm) [inline include/psa/crypto_compat.h:322]: I don't fully understand what's going on here. I think the point of the static assert is to cause an error if the macro is called with a number of arguments other than 1 or 3, but I don't fully understand how this works, and the error message in that case is incomprehensible.

Can't we do something simpler? Something like
```
#define PSA_AEAD_TAG_LENGTH_ARITY_CASE( arg1, arg2, arg3, arg_count, ... ) \
    ( arg_count == 1 ? PSA_AEAD_TAG_LENGTH_1( arg1 ) :                  \
      arg_count == 3 ? PSA_AEAD_TAG_LENGTH_3( arg1, arg2, arg3 ) :      \
      PSA_AEAD_TAG_LENGTH_called_with_0_or_2_arguments )
#define PSA_AEAD_TAG_LENGTH( ... )                                      \
    MBEDTLS_EXPAND( MBEDTLS_ARGS_LEN( __VA_ARGS__ ) <= 3 ?              \
                    PSA_AEAD_TAG_LENGTH_ARITY_CASE( __VA_ARGS__, 3, 2, 1 ) : \
                    PSA_AEAD_TAG_LENGTH_called_with_too_many_arguments )
```

Comment (bensze01) [reply include/psa/crypto_compat.h:322]: The main problem I see with your solution is `PSA_AEAD_TAG_LENGTH_called_with_too_many_arguments`.

I wanted to (reliably) cause a compilation error if someone used too many arguments. In your example, this problem would only be detectable at runtime - if `PSA_AEAD_TAG_LENGTH_called_with_too_many_arguments` is valid C code, it will compile without problem, and if it is not, it will error out even if the branch is not taken.

As for the rest, the way I've set it up the inappropriate macro never gets expanded. This reduces the total expanded size of the macro, and prevents stuff like `PSA_AEAD_TAG_LENGTH_3(foo, 3, 2)` being compiled in the untaken branch. 

I don't think this would cause any issues in our case, since these macros are only dealing with integers, but it would blow up if one of the macros was supposed to take something like a pointer or an array as an argument.

Comment (bensze01) [reply include/psa/crypto_compat.h:322]: You would also need to use `PSA_AEAD_TAG_LENGTH_ARITY_CASE( __VA_ARGS__, 3, 2, 1, 0 )` (padding with a 0), since an empty `...` argument is actually a gcc extension - ISO C does not allow it.

Basically, my code uses equivalent logic to your ARITY_CASE, just that I use PSA_AEAD_TAG_LENGTH_1_ARG and `PSA_AEAD_TAG_LENGTH_3_ARG` as the "counter" values instead of 3, 2, 1 and 0. The only values that are actually called are the first  `PSA_AEAD_TAG_LENGTH_3_ARG` (equivalent to 3 in your code), and the first PSA_AEAD_TAG_LENGTH_1_ARG (equivalent to 1 in your code), the other two are just the padding required to make this work, with values chosen to produce informative error messages if someone calls the macro with 2 arguments.

**Comment by bensze01** [comment]
The cause of the Travis error has been fixed upstream in development, it can be safely ignored.

File: include/psa/crypto_compat.h
Comment (gilles-peskine-arm) [reply include/psa/crypto_compat.h:322]: Yes, I see, thank you, what I attempted to do (leverage C expressions to detect errors) is fundamentally impossible.

I think that this use of the preprocessor is [too clever to be maintainable](https://www.quotes.net/quote/57502). I don't think it's worth the effort. I know you've already invested a lot of effort in making it work, but there will be maintenance cost: risk of encountering compiler bugs, time taken in debugging expressions involving these macros, supporting users who don't understand error messages...

I'm not going to veto it, but I will not spend the time to review it properly, which would require fully understanding it. As a gatekeeper, if two reviewers assert that they fully understand it and that they don't mind if [a future maintainer of the library who is a psychopath and knows their home address](https://github.com/ARMmbed/mbedtls-docs/blob/main/reviews/How%20to%20be%20an%20effective%20Mbed%20TLS%20reviewer.pdf) has to work on these macros, I'm ok with merging it.

As an architect, I recommend an alternative approach: keep the old macro as a deprecated `PSA_AEAD_TAG_LENGTH_1`, introduce the new macro as `PSA_AEAD_TAG_LENGTH` (and similarly for the other macros whose usage is changing). Document this in the changelog as an API change, listing the macros that need to be adapted. We do have a few precedents of doing not-fully-compatible API changes to align with the PSA specification.

Comment (ronald-cron-arm) [reply include/psa/crypto_compat.h:322]: I agree with the "As a architect" proposal. That's a much better way to go forward to me and the "keep the old macro as a deprecated PSA_AEAD_TAG_LENGTH_1" proposal mitigates significantly the negative impact on users: their code will not compile and they just have to add a "_1" to the modified macros to make it back to compile and work as before.

Otherwise, I don't feel I can be one of the reviewers that "don't mind if a future maintainer of the library who is a psychopath and knows their home address has to work on these macros".

**Comment by bensze01** [comment]
I've pushed a new version of the PR with the compatibility layer removed. I've kept the structure of the patch series otherwise intact, to facilitate range-diffing (use `git range-diff -w --creation-factor=80 development  e8fc54a 91a7ce9`)

The merge conflict should resolve itself once #4348 is merged to development.

**Comment by bensze01** [comment]
I've rebased this PR on top of development, to fix a merge conflict.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks overall good to me. I have a few suggestions for improvement.

The commit structure has simplified my review, thanks for that. I have a comment about it though. The CI does not test every commit and we don't have any plan to do so I think but the idea is that the vast majority of commit should pass CI at least the tests on the default configuration. Thus it is not really in the spirit I think to change the number of parameters of a macro in a commit and update its usage in a subsequent commit. I am not asking for a rewrite of the history of this PR, just mentioning it for future PRs.

Writing this I realize that maybe we ended-up in that situation because we eventually removed the macro-compatibility layer. I am leaving my comment anyway just in case and sorry if you were already aware of the development constraints at the commit level.

**Review by ronald-cron-arm — APPROVED** [review]
Thanks for the changes. LGTM.

**Comment by ronald-cron-arm** [comment]
@gabor-mezei-arm please review the latest version, thanks.

**Comment by bensze01** [comment]
Due to the change in merge strategy / merge of development_3.0 into development, we'll need a forward-port for this PR.

**Review by gabor-mezei-arm — APPROVED** [review]
LGTM

**Comment by ronald-cron-arm** [comment]
Please create the PR for `development` now.

**Pull request merged**
